### PR TITLE
Fixed #28821 -- Made QuerySet.bulk_create() work for multi-table inheritance when possible.

### PR DESCRIPTION
**Pull request by jdufresne**
ticket-28821

### CODE REVIEW

**Review by charettes — COMMENTED** [review]
Thanks for the patch

 I think the docs should mention that the current implementation of `bulk_create` requires one extra query per parent as one could assume that the implementation is smart enough to rely on `INSERT INTO child ... FROM SELECT id AS parent_ptr_id FROM (INSERT INTO parent ... RETURNING id)` which can be achieved in a single query.

**Comment by charettes** [comment]
Just to make it clear I think we should definitely move along with this patch even if it doesn't support single query MTI `bulk_create` for now as it addresses a relatively old limitation of this feature.

A follow up ticket that adds support for single query multi-table `INSERT`s and make `bulk_create` and `Model.save` rely on it when possible could then be created if one doesn't already exists.

**Comment by jdufresne** [comment]
Thanks for taking a look, Simon. I added "An additional SQL query is executed per concrete model in the inheritance hierarchy." to the `querysets.txt` docs. Please let me know if you think it could be improved, expanded, or tweaked.

**Review by charettes — COMMENTED** [review]
I left a comment regarding `Model.from_db` but outside out that it might be worth adding a test for direct multi table inheritance e.g.

```python
class A(models.Model):
    pass

class B(models.Model):
    pass

class C(A, B):
    pass
```

**Comment by jdufresne** [comment]
>  it might be worth adding a test for direct multi table inheritance e.g.

Thanks! This suggested test exposed a bug. I'm looking into resolving it.

**Comment by jdufresne** [comment]
I have implemented the suggested change and included the suggested test.

The precise test is slightly different than the suggested one due to avoid conflicts in PR fields which triggers a system check failure:

```
bulk_create.MultipleParents: (models.E005) The field 'id' from parent model 'bulk_create.nofields' clashes with the field 'id' from parent model 'bulk_create.twofields'.
```

File: tests/bulk_create/tests.py [hunk 1]
```diff
@@ -75,6 +77,122 @@ def test_multi_table_inheritance_unsupported(self):
                 ProxyMultiProxyCountry(name="Fillory", iso_two_letter="FL"),
             ])
 
+    @skipUnlessDBFeature('can_return_rows_from_bulk_insert')
+    def test_multi_table_inheritance(self):
+        pizzaria, = Pizzeria.objects.bulk_create([
```
Comment (felixxm) [inline tests/bulk_create/tests.py:82]:
```suggestion
        pizzeria, = Pizzeria.objects.bulk_create([
```

File: django/db/models/query.py [hunk 2]
```diff
@@ -510,17 +522,48 @@ def bulk_create(self, objs, batch_size=None, ignore_conflicts=False):
...
-                    obj_without_pk._state.adding = False
-                    obj_without_pk._state.db = self.db
+                if next_concrete_models:
+                    for next_concrete_model in next_concrete_models:
+                        # Recursively bulk insert parent objects.
+                        parent_objs = []
+                        for obj_without_pk in objs_without_pk:
+                            parent_obj_fields = []
+                            parent_obj_values = []
+                            for f in next_concrete_model._meta.concrete_fields:
```
Comment (felixxm) [inline django/db/models/query.py:532]: I think we should use `local_concrete_fields` :thinking: Also, the current solutions doesn't work with recursive parents, e.g.
```python
Pizzeria.objects.bulk_create([
    Pizzeria(name='Vegan Pizza House', rank=33),
])
```
crashes when we add the `Ranked` model:
```python
class Place(models.Model):
    name = models.CharField(max_length=100)

    class Meta:
        abstract = True


class Ranked(models.Model):
    rank = models.IntegerField(null=True)


class Restaurant(Ranked, Place):
    pass


class Pizzeria(Restaurant):
    pass
````
```

**Review by felixxm — COMMENTED** [review]
@jdufresne Thanks :+1:

**Comment by dralley** [comment]
I'd be glad to test this (or attempt to test it) on our codebase once it's ready.  We have a lot of multi-table inheritance.

**Comment by dralley** [comment]
@jdufresne Do you plan to pick this work back up?

**Comment by jdufresne** [comment]
@dralley I am no longer working on this. Please feel free to pick up the remaining work if you're motivated to do so. I will close this now to avoid any future confusion.

File: django/db/models/query.py (hunk 2 above)
Comment (dralley) [reply django/db/models/query.py:535]: Hi, I'm going to make an attempt at finishing this PR, but I'm not super familiar with Django innards.  Do you have any suggestions towards resolving this problem?

Comment (dralley) [reply django/db/models/query.py:535]: @felixxm ^

Comment (felixxm) [reply django/db/models/query.py:535]: @dralley The suggestion is already in the comment: _"I think we should use `local_concrete_fields`"_, but it's only a suggestion. I didn't check this.

Comment (dralley) [reply django/db/models/query.py:535]: Oh, I thought it was referring to two separate issues.

Alright then, I'll make a new PR and we can proceed from there

**Comment by mbrav** [comment]
ping

**Comment by dralley** [comment]
@mbrav  I haven't hada lot of time.  I do hope to pick it back up eventually.

**Pull request closed without merging**